### PR TITLE
Added Scanner to re.pyi

### PR DESCRIPTION
**Pull request by jonathanslenders**

### CODE REVIEW

File: stdlib/3/re.pyi [hunk 1]
```diff
@@ -153,3 +153,11 @@ def escape(string: AnyStr) -> AnyStr: ...
...
 def purge() -> None: ...
 def template(pattern: Union[AnyStr, Pattern[AnyStr]], flags: _FlagsType = ...) -> Pattern[AnyStr]: ...
+    
+class Scanner:
+    def __init__(
+        self,
+        lexicon: Sequence[Tuple[AnyStr, Callable[[Scanner, AnyStr], Any]]],
+        flags: _FlagsType = ...,
+    ) -> None: ...
+    def scan(self, string: AnyStr) -> Tuple[List[Any], AnyStr]: ...
```
Comment (srittau) [inline stdlib/3/re.pyi]: In the case of introducing another type var, `List[Any]` -> `List[_T]`.

**Comment by jonathanslenders** [comment]
Would this make sense?

```python
# undocumented

_T = TypeVar('_T', bound=AnyStr)
_T2 = TypeVar('_T2')

class Scanner(Generic[_T, _T2]):
    def __init__(
        self,
        lexicon: Sequence[Tuple[_T, Union[None, _T2, Callable[[Scanner, _T], Optional[_T2]]]]],
        flags: _FlagsType = ...,
    ) -> None: ...
    def scan(self, string: _T) -> Tuple[List[_T2], _T]: ...
```

Should we break it down in smaller pieces?

**Comment by srittau** [comment]
Looks good with two changes: The `# undocumented` should directly precede the class and `AnyStr` is already a type-var, so instead of `_T` you can use `AnyStr` directly (and don't need to list it in `Generic`).

**Comment by jonathanslenders** [comment]
So, something like this would do it?

```python
_T = TypeVar('_T')

# undocumented

class Scanner(Generic[_T]):
    def __init__(
        self,
        lexicon: Sequence[Tuple[AnyStr, Union[None, _T, Callable[[Scanner, AnyStr], Optional[_T]]]]],
        flags: _FlagsType = ...,
    ) -> None: ...
    def scan(self, string: AnyStr) -> Tuple[List[_T], AnyStr]: ...
```

What is the reason that ``AnyStr`` does not have to be listed in the Generic, while the other type variable should?

**Comment by JelleZijlstra** [comment]
> What is the reason that `AnyStr` does not have to be listed in the Generic, while the other type variable should?

In this case you have a generic method on a generic class. The class is generic over _T, and the method is generic over AnyStr. So for example, if you have a `Scanner[int]`, `scanner.scan("")` would return a `Tuple[List[int], str]`, and `scanner.scan(b"")` would return a `Tuple[List[int], bytes]`.

**Comment by srittau** [comment]
Closing for now, please reopen if you find time to work on this.

**Pull request closed without merging**